src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/Managed/Internal/QuicSocketContext.cs

```diff
@@ -0,0 +1,280 @@
+                }
+
+                connection = _connections![header.DestinationConnectionId];
+            }
+
+            reader.Seek(0);
+            bool connected = connection!.Connected;
+            connection.ReceiveData(reader, remoteEp, Timestamp.Now);
+            // TODO-RZ: handle failed connection attempts
+            if (!connected && connection!.Connected)
+            {
+                // new connection established
+                _newConnections!.TryWrite(connection!);
+            }
+
+            return connection;
+        }
+
+        protected override Task OnReceived(QuicReader reader, IPEndPoint sender)
+        {
+            var connection = DispatchPacket(reader, sender);
+            return connection != null ? SendAsync(connection) : Task.CompletedTask;
+        }
+
+        protected override async Task OnSignal()
+        {
+            foreach (var (_, connection) in _connections)
+            {
+                await SendAsync(connection);
+            }
+        }
+
+        protected override async Task OnTimeout()
+        {
+            // TODO-RZ: do timeout only on the one connection which needs it
+            long now = Timestamp.Now;
+            foreach (var (_, connection) in _connections)
+            {
+                connection.OnTimeout(now);
+                await SendAsync(connection);
+            }
+        }
+    }
+
+    internal sealed class QuicClientSocketContext : QuicSocketContext
+    {
+        private readonly ManagedQuicConnection _client;
+
+        internal QuicClientSocketContext(IPEndPoint localEndpoint, ManagedQuicConnection clientConnection)
+            : base(localEndpoint)
+        {
+            _client = clientConnection;
+        }
+
+        protected override Task OnReceived(QuicReader reader, IPEndPoint sender)
+        {
+            _client.ReceiveData(reader, sender, Timestamp.Now);
+            return SendAsync(_client);
+        }
+
+        protected override async Task OnSignal()
+        {
+             await SendAsync(_client);
+        }
+
+        protected override Task OnTimeout()
+        {
+            _client.OnTimeout(Timestamp.Now);
+            return SendAsync(_client);
+        }
+    }
+
+    /// <summary>
+    ///     Class responsible for serving a socket for QUIC connections.
+    /// </summary>
+    internal abstract class QuicSocketContext
+    {
+        protected readonly IPEndPoint _localEndpoint;
+        protected readonly CancellationTokenSource _socketTaskCts;
+
+        protected TaskCompletionSource<int> _signalTcs = new TaskCompletionSource<int>();
+
+        protected Task? _backgroundWorkerTask;
+
+        protected QuicReader _reader;
+        protected QuicWriter _writer;
+
+        private readonly Task _infiniteTimeoutTask = new TaskCompletionSource<int>().Task;
+        private Task _timeoutTask;
+        private long _currentTimeout = long.MaxValue;
+
+        private Task[] _waitingTasks = new Task[3];
```

Comment:
If you keep this array only because of the [`WaitAny`](https://github.com/rzikm/dotnet-runtime/blob/5f47e2cf8d99b178b2b1cab6923b5047407481c2/src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/Managed/Internal/QuicSocketContext.cs#L237), then you don't need to, `WaitAny` takes `params`.